---
title: "Bulk RNA-seq Data Set - GTEx - v8"
author: "Regina H. Reynolds"
output: 
  html_document:
    theme: paper
    highlight: kate
    df_print: paged
    toc: true
    toc_float: true
    number_sections: true
---

```{r setup, include=FALSE}

library(here)
library(tidyverse)

knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

> Code to generate this is originally from the paper [Bryois et al.](https://www.nature.com/articles/s41588-020-0610-9), available at: https://github.com/jbryois/scRNA_disease/blob/master/Code_Paper/Code_GTEx/get_GTEx_input.md. The original code has been modified below to: (1) Include all genes, as opposed to including only protein-coding genes; (2) remove brain cortex and brain cerebellum to reduce redundancy; (3) not to add gene co-ordinates; (3) to make some column names more human-readable i.e. in the modified code below, what was "Lvl5" ("SMTSD" in GTEx) is now referred to as "Tissue", and what was "Tissue" ("SMTS" in GTEx) is referred to as "Organ".

# Load Data

## Load GTEx dataset

The GTEx data (Version 8) was downloaded from [here](https://gtexportal.org)

```{r,message=FALSE}
file <- here::here("inst", "extdata", "GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct.gz")
exp <- 
  read_tsv(file, skip=2) %>% 
  dplyr::rename(gene_id = Name)

```

## Collapse expression per tissue

The gtex_v8_tissues.tsv contains the tissues to keep (N>100), as well as info about which organ each tissue belongs to. We want to keep the different Brain tissues and not collapsed at the organ level for the Brain. So we add the different Brain tissues to the organ level column.

```{r}
tissues <-
  read_delim(here::here("inst", "extdata", "gtex_v8_tissues.tsv"), delim = " ") %>% 
  dplyr::rename(Tissue = SMTSD, Organ = SMTS) %>% 
  dplyr::mutate(Organ = case_when(str_detect("Brain", Organ) ~ Tissue,
                                    TRUE ~ Organ))


```

We tidy the GTEx gene expression data and join with the tissue information file

```{r}

exp <- 
  exp %>% 
  tidyr::gather(key = Tissue, value=Expr,-gene_id,-Description) %>% 
  as_tibble()

exp <- 
  exp %>% 
  inner_join(tissues)
```

We then drop tissues with: 
    - Less than 100 samples
    - Testis (gene expression outlier)
    - Non natural tissues (e.g. EBV-transformed lymphocytes and cultured fibroblasts)
    - Brain cortex and Brain cerebellum (using instead the more specific anterior cingulate cortex, frontal cortex and cerebellar hemisphere) to reduce redundancy across brain regions.
We then take the average expression for the different tissues that belong to the same organ.

```{r}

tissues_to_exclude <- c("Testis", "Brain - Cerebellum", "Brain - Cortex", "Cells - EBV-transformed lymphocytes", "Cells - Cultured fibroblasts")

exp_per_organ <- 
  exp %>% 
  dplyr::filter(Nsample >= 100, !Tissue %in% tissues_to_exclude) %>%
  dplyr::group_by(Organ,gene_id,Description) %>% 
  dplyr::summarise(mean_expr = mean(Expr))

```

Spread the data.

```{r}
exp <- 
  exp_per_organ %>% 
  tidyr::spread(Organ, mean_expr) %>% 
  dplyr::ungroup()

```

## Scale to 1M TPM

```{r}

exp_scaled <- apply(exp[-c(1,2)],2,function(x) x*1e6/sum(x))
exp <- 
  cbind(exp[c(1,2)],exp_scaled) %>% 
  as_tibble()

```

Only keep genes with a unique name

```{R, load data 3,message=FALSE, warning=FALSE}
exp <- 
  exp %>% 
  dplyr::add_count(gene_id) %>% 
  dplyr::filter(n==1) %>%
  dplyr::select(-n) %>%
  dplyr::mutate(gene_id = gsub("\\..+","",gene_id)) %>%
  tidyr::gather(key = Organ, value=Expr,-gene_id,-Description) %>% 
  as_tibble()
```

# QC

## Remove not expressed genes

```{R, rm lowQ 2,message=F}

exp_tissue <- 
  exp %>% 
  dplyr::rename(mean_expr=Expr,Gene=gene_id)

not_expressed <- 
  exp_tissue %>% 
  dplyr::group_by(Gene) %>% 
  dplyr::summarise(total_sum=sum(mean_expr)) %>% 
  dplyr::filter(total_sum==0) %>% 
  dplyr::select(Gene) %>% 
  unique() 

exp_tissue <- 
  exp_tissue %>% 
  dplyr::filter(!Gene %in% not_expressed$Gene)

```

# Specificity Calculation

The specifitiy is defined as the proportion of total expression performed by the cell type of interest (x/sum(x)).

```{R, specificity 5,message=F}
exp_tissue <- 
  exp_tissue %>% 
  dplyr::group_by(Gene) %>% 
  dplyr::mutate(specificity= mean_expr/sum(mean_expr)) %>% 
  dplyr::ungroup()

saveRDS(exp_tissue, "../specificity_df/GTEx_v8.Rds")
```
