---
title: "Tutorial: querying genes across specificity matrices"
author: 
  - name: Regina H. Reynolds
    affiliation:
    - UCL
    email: regina.reynolds.16@ucl.ac.uk
output: 
  BiocStyle::html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
package: "`r pkg_ver('MarkerGenes')`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r setup, include = FALSE}
library(DT)
library(here)
library(MarkerGenes)
library(readxl)
library(tidyverse)

theme_rhr <-  
  ggplot2::theme_bw(base_family = "Helvetica") + 
  ggplot2::theme(
    panel.grid.major.x = ggplot2::element_blank(),
        legend.position = "right",
        strip.text = ggplot2::element_text(size = 7),
        axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1, vjust = 0.5),
        axis.text.y = ggplot2::element_text(size = 7),
        axis.title.y = ggplot2::element_text(vjust = 0.6),
        axis.title = ggplot2::element_text(size = 10),
        panel.spacing = ggplot2::unit(0.1, "lines")
    )

knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    crop = NULL ## Related to https://stat.ethz.ch/pipermail/bioc-devel/2020-April/016656.html
)
```

# Aim
- Quick tutorial of `MarkerGenes::query_gene_ctd()` function, which allows users to query a gene or several genes within several specificity matrices.

# Mock example
- We will use the human gene, SNCA, as our mock example.
- We are only interested in SNCA's specificity in human datasets, thus we will query the following datasets:

```{r read-metadata}

readxl::read_excel(
  path = here::here("metadata", "dataset_metadata.xlsx"),
  sheet = "Data"
  ) %>% 
  dplyr::filter(species == "human", data_type == "SI_matrix") %>% 
  DT::datatable(rownames = FALSE,
                options = list(scrollX = TRUE),
                class = 'white-space: nowrap')

```

- As AIBS2018_MTG is the more up to data of the AIBS datasets, we will use this, together with Habib2017_DroNc_Human. In addition, we will be using some of our own in-house single-nuclear RNA-sequencing from the anterior cingulate cortex of control individuals (n = 7).
    - It is worth noting that there are key differences in methodology between the AIBS dataset and the two remaining (i.e. Habib and our own in-house snRNA). While both sequence single nuclei, AIBS uses the SMART-Seq v4 kit, which permits sequence of full-length transcripts, while Habib and our in-house data use a 3'-tagging technology i.e. not full-length transcripts sequenced. This may have an impact on gene specificity.
- Let's go ahead and load in these specificity matrices and query them for SNCA specificity.
- As we are only interested in SNCA's specificity in overarching cell types, let's limit this search to level 1 cell types.

```{r specificity-plot, fig.cap = "**Figure**: Plot of SNCA specificity value across three single-nuclear RNA-seq datasets. exPFC=glutamatergic neurons from the PFC, exCA1/3=pyramidal neurons from the Hip CA region, GABA=GABAergic interneurons, exDG=granule neurons from the Hip dentate gyrus region, ASC=astrocytes, NSC=neuronal stem cells, MG=microglia, ODC=oligodendrocytes, OPC=oligodendrocyte precursor cells, NSC=neuronal stem cells, SMC=smooth muscle cells, END= endothelial cells."}

# Load specificity matrices
load(here::here("specificity_matrices", "AIBS2018_MTG.rda"))
load(here::here("specificity_matrices", "Habib2017_DroNc_Human.rda"))
load(here::here("specificity_matrices", "Agarwal2020_SNIG.rda"))

specificity <- 
  MarkerGenes::query_gene_ctd(
    genes = c("SNCA"),
    ctd_AIBS2018, ctd_DRONC_human, ctd_Agarwal2020_SNIG,
    celltypeLevel = 1, 
    median_included = F,
    genelistSpecies = "human", 
    ctdSpecies = "human"
  )

# Plot
specificity %>% 
  ggplot2::ggplot(
    ggplot2::aes(
      x = MarkerGenes::reorder_within(
        x = CellType, 
        by = Specificity, 
        within = Study, 
        fun = median, 
        desc = T
      ), 
      y = Specificity)) + 
  ggplot2::geom_col() +
  MarkerGenes::scale_x_reordered() +
  ggplot2::facet_wrap(ggplot2::vars(Study), scales = "free_x") + 
  ggplot2::labs(x = "CellType") +
  theme_rhr

```

- For all three datasets, SNCA appears to have its highest specificity in some form of excitatory neuron. 
- Thereafter, the ordering of cell types is a little more ambiguous, and changes slightly dependent on the dataset. 
