<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tutorial: querying genes across specificity matrices â€¢ MarkerGenes</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Tutorial: querying genes across specificity matrices">
<meta property="og:description" content="MarkerGenes">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">MarkerGenes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../articles/MarkerGenes.html">Get started</a>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/articles/dataset_metadata.html">Dataset metadata</a>
    </li>
    <li>
      <a href="../../articles/articles/query_gene_ctd_tutorial.html">Tutorial: querying genes across specificity matrices</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="query_gene_ctd_tutorial_files/htmlwidgets-1.5.4/htmlwidgets.js"></script><link href="query_gene_ctd_tutorial_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="query_gene_ctd_tutorial_files/datatables-binding-0.19/datatables.js"></script><script src="query_gene_ctd_tutorial_files/jquery-3.6.0/jquery-3.6.0.min.js"></script><link href="query_gene_ctd_tutorial_files/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="query_gene_ctd_tutorial_files/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="query_gene_ctd_tutorial_files/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="query_gene_ctd_tutorial_files/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="query_gene_ctd_tutorial_files/crosstalk-1.1.1/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tutorial: querying genes across specificity matrices</h1>
                        <h4 class="author">Regina H. Reynolds</h4>
            <address class="author_afil">
      UCL<br><a class="author_email" href="mailto:#"></a><a href="mailto:regina.reynolds.16@ucl.ac.uk">regina.reynolds.16@ucl.ac.uk</a>
      </address>
                  
            <h4 class="date">6 April 2022</h4>
      
      
      <div class="hidden name"><code>query_gene_ctd_tutorial.Rmd</code></div>

    </div>

    
    
<div id="aim" class="section level1">
<h1 class="hasAnchor">
<a href="#aim" class="anchor"></a>Aim</h1>
<ul>
<li>Quick tutorial of <code><a href="../../reference/query_gene_ctd.html">MarkerGenes::query_gene_ctd()</a></code> function, which allows users to query a gene or several genes within several specificity matrices.</li>
</ul>
</div>
<div id="mock-example" class="section level1">
<h1 class="hasAnchor">
<a href="#mock-example" class="anchor"></a>Mock example</h1>
<ul>
<li>We will use the human gene, SNCA, as our mock example.</li>
<li>We are only interested in SNCA's specificity in human datasets, thus we will query the following datasets:</li>
</ul>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="fu">readxl</span><span class="fu">::</span><span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html">read_excel</a></span><span class="op">(</span>
  path <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"metadata"</span>, <span class="st">"dataset_metadata.xlsx"</span><span class="op">)</span>,
  sheet <span class="op">=</span> <span class="st">"Data"</span>
  <span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">species</span> <span class="op">==</span> <span class="st">"human"</span>, <span class="va">data_type</span> <span class="op">==</span> <span class="st">"SI_matrix"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html">datatable</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="cn">FALSE</span>,
                options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>scrollX <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,
                class <span class="op">=</span> <span class="st">'white-space: nowrap'</span><span class="op">)</span></code></pre></div>
<div id="htmlwidget-4fd8e32e164411c29855" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-4fd8e32e164411c29855">{"x":{"filter":"none","vertical":false,"data":[["Habib2017_DroNc_Human","AIBS2018_MTG","Agarwal2020_SNIG","Agarwal2020_CRTX","Lake2018_FrontalCortexOnly","Lake2018_VisualCortexOnly"],["Habib","Unpublished","Agarwal","Agarwal","Lake","Lake"],["2017","NA","2020","2020","2018","2018"],["https://www.nature.com/articles/nmeth.4407","NA","https://www.nature.com/articles/s41467-020-17876-0","https://www.nature.com/articles/s41467-020-17876-0","https://www.ncbi.nlm.nih.gov/pubmed/29227469","https://www.ncbi.nlm.nih.gov/pubmed/29227469"],["human","human","human","human","human","human"],["hippocampus;pre_frontal_cortex","middle_temporal_gyrus","substantia_nigra","cortex","frontal_cortex_BA6_BA10","visual_cortex_BA17"],["control;post_mortem","control;post_mortem","control;post_mortem","control;post_mortem","control;post_mortem","control;post_mortem"],["single_nuclei","single_nuclei","single_nuclei","single_nuclei","single_nuclei","single_nuclei"],["5","8","5","5","4","5"],["19550","15928","6105","12015","10319","19368"],["10","7","7","6","8","8"],["15","75","10","23","24","28"],["https://github.com/NathanSkene/MAGMA_Celltyping/tree/master/data","https://portal.brain-map.org/atlases-and-data/rnaseq/human-mtg-smart-seq","https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE140231","https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE140231","https://github.com/NathanSkene/MAGMA_Celltyping","https://github.com/NathanSkene/MAGMA_Celltyping"],["ctd_DRONC_human.rda","human_MTG_gene_expression_matrices_2018-06-14.zip","GSE140231_RAW.tar","GSE140231_RAW.tar","ctd_BlueLake2018_FrontalCortexOnly.Rda","ctd_BlueLake2018_VisualCortexOnly.Rda"],["SI_matrix","SI_matrix","SI_matrix","SI_matrix","SI_matrix","SI_matrix"]],"container":"<table class=\"white-space: nowrap\">\n  <thead>\n    <tr>\n      <th>resource_name<\/th>\n      <th>first_author<\/th>\n      <th>publication_year<\/th>\n      <th>publication_link<\/th>\n      <th>species<\/th>\n      <th>tissue_region<\/th>\n      <th>model<\/th>\n      <th>single_or_bulk<\/th>\n      <th>number_of_samples_donors<\/th>\n      <th>number_of_cells<\/th>\n      <th>number_of_celltypes_identified_level1<\/th>\n      <th>number_of_celltypes_identified_level2<\/th>\n      <th>download_link<\/th>\n      <th>original_filename<\/th>\n      <th>data_type<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><ul>
<li>As AIBS2018_MTG is the more up to data of the AIBS datasets, we will use this, together with Habib2017_DroNc_Human. In addition, we will be using some of our own in-house single-nuclear RNA-sequencing from the anterior cingulate cortex of control individuals (n = 7).
<ul>
<li>It is worth noting that there are key differences in methodology between the AIBS dataset and the two remaining (i.e. Habib and our own in-house snRNA). While both sequence single nuclei, AIBS uses the SMART-Seq v4 kit, which permits sequence of full-length transcripts, while Habib and our in-house data use a 3'-tagging technology i.e. not full-length transcripts sequenced. This may have an impact on gene specificity.</li>
</ul>
</li>
<li>Let's go ahead and load in these specificity matrices and query them for SNCA specificity.</li>
<li>As we are only interested in SNCA's specificity in overarching cell types, let's limit this search to level 1 cell types.</li>
</ul>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># Load specificity matrices</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"specificity_matrices"</span>, <span class="st">"AIBS2018_MTG.rda"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"specificity_matrices"</span>, <span class="st">"Habib2017_DroNc_Human.rda"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"specificity_matrices"</span>, <span class="st">"Agarwal2020_SNIG.rda"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; DEPRECATION NOTICE:</span>
<span class="co">#&gt; This version of the EWCE package is now deprecated. Please install EWCE from Bioconductor (3.13) with</span>
<span class="co">#&gt; R &gt;= 4.1. See the new repository for more information (https://github.com/NathanSkene/EWCE).One thing </span>
<span class="co">#&gt; to note is that all functions which previously used dot '.' notations to separate words have been updated</span>
<span class="co">#&gt; to underscores '_' (e.g, `bootstrap.enrichment.test()` is now `bootstrap_enrichment_test()`). The package</span>
<span class="co">#&gt; vignette website (https://nathanskene.github.io/EWCE/articles/EWCE.html) has been updated to this new</span>
<span class="co">#&gt; version of EWCE and gives further details.</span>

<span class="va">specificity</span> <span class="op">&lt;-</span> 
  <span class="fu">MarkerGenes</span><span class="fu">::</span><span class="fu"><a href="../../reference/query_gene_ctd.html">query_gene_ctd</a></span><span class="op">(</span>
    genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"SNCA"</span><span class="op">)</span>,
    <span class="va">ctd_AIBS2018</span>, <span class="va">ctd_DRONC_human</span>, <span class="va">ctd_Agarwal2020_SNIG</span>,
    celltypeLevel <span class="op">=</span> <span class="fl">1</span>, 
    median_included <span class="op">=</span> <span class="cn">F</span>,
    genelistSpecies <span class="op">=</span> <span class="st">"human"</span>, 
    ctdSpecies <span class="op">=</span> <span class="st">"human"</span>
  <span class="op">)</span>

<span class="co"># Plot</span>
<span class="va">specificity</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>
      x <span class="op">=</span> <span class="fu">MarkerGenes</span><span class="fu">::</span><span class="fu"><a href="../../reference/reorder_within.html">reorder_within</a></span><span class="op">(</span>
        x <span class="op">=</span> <span class="va">CellType</span>, 
        by <span class="op">=</span> <span class="va">Specificity</span>, 
        within <span class="op">=</span> <span class="va">Study</span>, 
        fun <span class="op">=</span> <span class="va">median</span>, 
        desc <span class="op">=</span> <span class="cn">T</span>
      <span class="op">)</span>, 
      y <span class="op">=</span> <span class="va">Specificity</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_col</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">MarkerGenes</span><span class="fu">::</span><span class="fu"><a href="../../reference/reorder_within.html">scale_x_reordered</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html">vars</a></span><span class="op">(</span><span class="va">Study</span><span class="op">)</span>, scales <span class="op">=</span> <span class="st">"free_x"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"CellType"</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">theme_rhr</span></code></pre></div>
<div class="figure">
<img src="query_gene_ctd_tutorial_files/figure-html/specificity-plot-1.png" alt="**Figure**: Plot of SNCA specificity value across three single-nuclear RNA-seq datasets. exPFC=glutamatergic neurons from the PFC, exCA1/3=pyramidal neurons from the Hip CA region, GABA=GABAergic interneurons, exDG=granule neurons from the Hip dentate gyrus region, ASC=astrocytes, NSC=neuronal stem cells, MG=microglia, ODC=oligodendrocytes, OPC=oligodendrocyte precursor cells, NSC=neuronal stem cells, SMC=smooth muscle cells, END= endothelial cells." width="700"><p class="caption">
<strong>Figure</strong>: Plot of SNCA specificity value across three single-nuclear RNA-seq datasets. exPFC=glutamatergic neurons from the PFC, exCA1/3=pyramidal neurons from the Hip CA region, GABA=GABAergic interneurons, exDG=granule neurons from the Hip dentate gyrus region, ASC=astrocytes, NSC=neuronal stem cells, MG=microglia, ODC=oligodendrocytes, OPC=oligodendrocyte precursor cells, NSC=neuronal stem cells, SMC=smooth muscle cells, END= endothelial cells.
</p>
</div>
<ul>
<li>For all three datasets, SNCA appears to have its highest specificity in some form of excitatory neuron.</li>
<li>Thereafter, the ordering of cell types is a little more ambiguous, and changes slightly dependent on the dataset.</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Regina H. Reynolds.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
